### PR TITLE
feat: add blocks bulk-load command for historic plugin.

### LINKED ISSUE #2955: CLI: backfill historical WRBs by pushing to a Block Node
Part of epic #2954 — _Distribute historical & live WRBs and TSS to Tier-1 Block Nodes_.

**Prereq:** none. (Full historical-WRB serving also needs T4+T5, but T1 is independently buildable/testable.)

## Story Form

As a node operator running the CLI conversion server,
I want a CLI command that reads the existing wrapped-block zips and pushes each block to a Block Node's ingestion endpoint,
So that the special-purpose BN is seeded with the full historical WRB set without restructuring storage.

## Technical Notes

- New/extended `blocks` subcommand (e.g. `blocks push`) reading the wrapped-block zip storage (10k blocks/zip) and publishing **one block at a time**, in order, over gRPC.
- Reuse the publish client used by `networkCapacity` client mode / simulator `PublishStreamGrpcClient`; server side is the standard `stream-publisher` ingestion.
- Resumable: start from the BN's latest stored block (query it) so a restart continues, not restarts.
- Options: source dir, target endpoint, start/resume block, rate limit. Emit progress + last-pushed block metrics/logs. Handle BN-unavailable with backoff; never drop a block.

## Design

`docs/design/wrb-streaming/sp-wrb-bn-design.md`

### PR DESCRIPTION
**Pull request by rockysingh**
## Description

Implements CLI `blocks bulk-load` command to copy wrapped block zips from CLI output into Block Node historic storage for the BlockFileHistoricPlugin.

Closes #2955

## Features

- **Bulk copy**: Copies wrapped block zip files from source directory to Block Node historic storage directory
- **Resume support**: Tracks progress in `historic-plugin-bulk-load-state.json` and can resume from interruption
- **Filtering**: `--start-block` option to only copy blocks >= specified number
- **Dry-run mode**: Preview what would be copied without making changes
- **Progress tracking**: Shows copy progress every 10 files and saves state periodically

## Implementation

- `BulkLoadBlocksCommand.java`: Main command implementation with resume functionality
- `BulkLoadBlocksCommandTest.java`: Comprehensive test suite with 16 tests covering:
  - Command line interface (3 tests)
  - File copying (4 tests)
  - State file management (3 tests)
  - Start block filtering (2 tests)
  - Dry run mode (1 test)
  - Edge cases (3 tests)

## Usage

```bash
# Basic bulk load
blocks bulk-load --source-dir /path/to/cli/output --dest-dir /opt/hiero/block-node/data/historic

# Resume from previous run
blocks bulk-load --source-dir /path/to/cli/output --dest-dir /opt/hiero/block-node/data/historic --resume

# Copy only blocks >= 1000
blocks bulk-load --source-dir /path/to/cli/output --dest-dir /opt/hiero/block-node/data/historic --start-block 1000

# Dry run to preview
blocks bulk-load --source-dir /path/to/cli/output --dest-dir /opt/hiero/block-node/data/historic --dry-run
```

### CODE REVIEW

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+            return;
+        }
+
+        try {
+            // Extract the first block number from the zip file name
+            String numStr = fileName.substring(0, fileName.length() - 5); // Remove "s.zip"
+            long zipFirstBlock = Long.parseLong(numStr);
+
+            // Each zip contains 10,000 blocks
+            long zipLastBlock = zipFirstBlock + 10_000 - 1;
```
Comment (Nana-EC) [inline tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java]: Use config to allow for customization
```suggestion
            long zipLastBlock = zipFirstBlock + BlockZipsUtilities.DEFAULT_BLOCKS_PER_ZIP - 1;
```

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+        try {
+            // Extract the first block number from the zip file name
+            // e.g., "10000s.zip" -> 10000
+            String numStr = fileName.substring(0, fileName.length() - 5); // Remove "s.zip"
+            long zipFirstBlock = Long.parseLong(numStr);
+
+            // Determine the last block in this zip
+            // Assume powersOfTen=4 (10,000 blocks per zip) as that's the default
+            long blocksPerZip = 10_000;
+            long zipLastBlock = zipFirstBlock + blocksPerZip - 1;
```
Comment (Nana-EC) [inline tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java]: use config
```suggestion
            long zipLastBlock = zipFirstBlock + BlockZipsUtilities.DEFAULT_BLOCKS_PER_ZIP - 1;
```

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+                    copiedBytes.set(state.totalBytesCopied);
+                    progress.advance(attrs.size());
+
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                    // Skip staging, links, and zipwork directories if they exist in source
+                    String dirName = dir.getFileName().toString();
```
Comment (Nana-EC) [inline tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java]: Missing a NPE check like you for for scanWorkToDo
```suggestion
                    String dirName = dir.getFileName() == null ? "" : dir.getFileName().toString();
```

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+                if (!file.toString().endsWith(".zip")) {
+                    return FileVisitResult.CONTINUE;
+                }
+                Path rel = source.relativize(file);
+                if (state.copiedFiles.contains(rel.toString())) {
```
Comment (Nana-EC) [inline tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java:519]: On a resume run, every file visit scans the entire list. This will be large for mainnet no?
Should we explore an improvement to speed things up?
Might be worth exploring if copiedFiles should be HashSet.

**Review by Nana-EC — DISMISSED** [review]
LG, some suggested improvements and consideration that you can address in follow up if needed.

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java (hunk 2 above)
Comment (rockysingh) [reply tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java]: Done in a6f0f226 — replaced the hardcoded `10_000` in `shouldSkipBasedOnBlockNumber` with `BlockZipsUtilities.DEFAULT_BLOCKS_PER_ZIP`, matching the fix applied to `updateBlockNumber` earlier. Also added the missing import (the earlier Apply Suggestion didn't add it, which was breaking the build).

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java (hunk 1 above)
Comment (rockysingh) [reply tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java]: Same fix as the line 420 thread — done in a6f0f226.

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java (hunk 4 above)
Comment (rockysingh) [reply tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java:519]: Good catch on the O(n²) resume scan. Changed `copiedFiles` from `ArrayList<String>` to `HashSet<String>` in a6f0f226 so `.contains()` is O(1) instead of O(n). Jackson serializes `Set` as a JSON array, so on-disk state files remain backward-compatible (existing state files deserialize fine into a HashSet).

**Review by Nana-EC — DISMISSED** [review]

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java [hunk 5]
```diff
@@ -0,0 +1,633 @@
+// SPDX-License-Identifier: Apache-2.0
+package org.hiero.block.tools.blocks;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
```
Comment (jsync-swirlds) [inline tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java]: Instead of Jackson; can we use the PBJ JSON support for this?

**Review by jsync-swirlds — COMMENTED** [review]
Might be good to stick with PBJ if we can for JSON processing.

File: tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java (hunk 5 above)
Comment (rockysingh) [reply tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java]: Yep will get that in.

Comment (rockysingh) [reply tools-and-tests/tools/src/main/java/org/hiero/block/tools/blocks/BulkLoadBlocksCommand.java]: Done in c92d6bab — added `internal/bulk_load_state.proto` defining `BulkLoadState` and migrated `BulkLoadBlocksCommand` to `BulkLoadState.JSON.parse / .write` (via PBJ's `ReadableStreamingData / WritableStreamingData`). Kept the mutable `LoadState` POJO in-memory so `copiedFiles` stays a `HashSet` for O(1) lookups during the walk; convert to/from the PBJ message at the I/O boundary. Also dropped Jackson from `mainModuleInfo` and migrated the tests to the same PBJ JSON path. PBJ emits camelCase keys so on-disk state files written by previous versions remain readable.

**Review by jsync-swirlds — APPROVED** [review]

**Review by Nana-EC — APPROVED** [review]
